File: graph_search/200_number_of_islands.cc
#include <queue>
#include <vector>

// 2024/01/10: 自己不看答案实现.
int NumIslands(const std::vector<std::vector<char>>& grid) {
  int m = grid.size();
  int n = grid[0].size();
  std::vector<std::vector<int>> is_visited(m, std::vector<int>(n, 0));
  int res = 0;
  std::vector<std::vector<int>> directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
  for (int i = 0; i < m; ++i) {
    for (int j = 0; j < n; ++j) {
      if (grid[i][j] == '1' && !is_visited[i][j]) {
        ++res;
        std::queue<std::pair<int, int>> q;
        q.push(std::make_pair(i, j));
        while (!q.empty()) {
          std::pair<int, int> temp = q.front();
          q.pop();
          for (const auto& direction : directions) {
            int new_row = temp.first + direction[0];
            int new_col = temp.second + direction[1];
            if (new_row >= 0 && new_row < m && new_col >= 0 && new_col < n &&
                !is_visited[new_row][new_col] &&
                grid[new_row][new_col] == '1') {
              q.push(std::make_pair(new_row, new_col));
              is_visited[new_row][new_col] = 1;
            }
          }
        }
      }
    }
  }
  return res;
}

File: graph/200_number_of_islands.py
from typing import List


# 2026/05/20: 自己不看答案实现.
def dfs(
    grid: List[List[str]], row: int, col: int, is_visited: List[List[bool]]
) -> None:
    m, n = len(grid), len(grid[0])
    if row < 0 or row >= m or col < 0 or col >= n:
        return

    if grid[row][col] == "0" or is_visited[row][col]:
        return

    is_visited[row][col] = True
    dfs(grid, row - 1, col, is_visited)
    dfs(grid, row + 1, col, is_visited)
    dfs(grid, row, col - 1, is_visited)
    dfs(grid, row, col + 1, is_visited)


def num_islands(grid: List[List[str]]) -> int:
    m, n = len(grid), len(grid[0])
    is_visited = [[False for _ in range(n)] for _ in range(m)]
    res = 0

    for i in range(m):
        for j in range(n):
            if grid[i][j] == "1" and not is_visited[i][j]:
                res += 1
                dfs(grid, i, j, is_visited)

    return res


grid = [
    ["1", "1", "1", "1", "0"],
    ["1", "1", "0", "1", "0"],
    ["1", "1", "0", "0", "0"],
    ["0", "0", "0", "0", "0"],
]
print(num_islands(grid))
